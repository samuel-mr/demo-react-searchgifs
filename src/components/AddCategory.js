import React, { useState } from 'react'
import PropTypes from 'prop-types'

export const AddCategory = ({ setCategory }) => {
    const [value, setValue] = useState('')

    const onInputChange = (e) => {
        setValue(e.target.value)
    }

    const handleSubmit = (e) => {
        e.preventDefault()
        if (value.trim().length > 2) {
            setCategory(currentValue => [value, ...currentValue])
            setValue('')
        }
    }

    return (
        <form onSubmit={handleSubmit}>
            <p>{value}</p>
            <input type='text' value={value} onChange={onInputChange}></input>

        </ form >
    )

}

AddCategory.propTypes = {
    setCategory: PropTypes.func.isRequired
}