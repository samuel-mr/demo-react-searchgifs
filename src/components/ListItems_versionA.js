import React, { useState, useEffect } from 'react'
import { ListItemImage } from './ListItemImage'
import { getGifs } from '../helpers/getGifs'

export const Listitem = ({ category }) => {

    const [images, setImages] = useState([])

    useEffect(() => {
        getGifs(category)
            .then(imgs => setImages(imgs))
    }, [category])


    return (
        <>
            <h1>{category}</h1>
            <div className='card-grid'>
                {
                    images.map(item => (
                        <ListItemImage
                            key={item.id}
                            {...item}
                        />
                    ))
                }
            </div>
        </>
    )
}