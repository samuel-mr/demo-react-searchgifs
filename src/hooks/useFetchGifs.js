import { useState, useEffect } from "react"
import { getGifs } from '../helpers/getGifs'

export const useFetchGifs = (category) => {
    const [state, setState] = useState({
        data: [],
        loading: true
    })

    useEffect(() => {                                                                
        getGifs(category)
            .then(imgs => {
                setTimeout(() => { //simulate delay
                    setState({
                        data: imgs,
                        loading: false
                    })
                }, 500) //medio minuto
            })
    }, [category])

    return state
}