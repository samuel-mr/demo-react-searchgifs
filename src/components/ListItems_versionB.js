import React from 'react'
import { ListItemImage } from './ListItemImage'
import { useFetchGifs } from '../hooks/useFetchGifs'

export const ListItems_versionB = ({ category }) => {

    const { data:images, loading } = useFetchGifs(category)


    return (
        <>
            <h1 className='animate__animated animate__slideInDown'>{category}</h1>
            {loading && <div className='animate__animated animate__flash'>Cargando... </div>}
            <div className='card-grid'>
                {
                    images.map(item => (
                        <ListItemImage
                            key={item.id}
                            {...item}
                        />
                    ))
                }
            </div>
        </>
    )
}