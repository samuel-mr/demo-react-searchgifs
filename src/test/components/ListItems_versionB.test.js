import React from 'react'
import { shallow } from 'enzyme'
import { ListItems_versionB } from '../../components/ListItems_versionB'
import { useFetchGifs } from '../../hooks/useFetchGifs'
jest.mock('../../hooks/useFetchGifs')

describe('testing ListItem_versionB', () => {
    test('Debe ser igual al snapshot', () => {
        const CATEGORIA = 'simple'

        // Sobreescribiendo la llamada a 'useFetchGifs'
        useFetchGifs.mockReturnValue({
            data: [],
            loading: true
        })

        const wrapper = shallow(<ListItems_versionB category={CATEGORIA} />)
        expect(wrapper).toMatchSnapshot()
    })

    test('Debe mostrar items cuando se cargan las imagenes', () => {
        const CATEGORIA = 'simple'

        const DATA = [{
            id: 123,
            url: 'http:direcion.com',
            title: 'demo'
        },{
            id: 1235,
            url: 'http:direcion.com',
            title: 'demo'
        }]

        // 1. Sobreescribiendo el valor de retorno de la función 'useFetchGifs'
        useFetchGifs.mockReturnValue({
            data: DATA,
            loading: false
        })

        const wrapper = shallow(<ListItems_versionB category={CATEGORIA} />)

        // 2. Verificar que el contenido sea igual al snapshot
        expect(wrapper).toMatchSnapshot()

        // 3. Verificar que exista al menos un ListItemImage (ya que tengo resultados)
        expect(wrapper.find('ListItemImage').exists()).toBeTruthy()
        // 4. Verificar que la cantidad de Items sea igual al tamaño del arreglo
        expect(wrapper.find('ListItemImage').length).toBe(DATA.length)

        // 5. Probar que el parrafo no exista (ya que el loading es false)
        expect(wrapper.find('p').exists()).toBeFalsy()

    })
})