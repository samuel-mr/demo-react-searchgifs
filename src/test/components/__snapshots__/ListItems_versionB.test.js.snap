// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing ListItem_versionB Debe mostrar items cuando se cargan las imagenes 1`] = `
<Fragment>
  <h1
    className="animate__animated animate__slideInDown"
  >
    simple
  </h1>
  <div
    className="card-grid"
  >
    <ListItemImage
      id={123}
      key="123"
      title="demo"
      url="http:direcion.com"
    />
    <ListItemImage
      id={1235}
      key="1235"
      title="demo"
      url="http:direcion.com"
    />
  </div>
</Fragment>
`;

exports[`testing ListItem_versionB Debe ser igual al snapshot 1`] = `
<Fragment>
  <h1
    className="animate__animated animate__slideInDown"
  >
    simple
  </h1>
  <div
    className="animate__animated animate__flash"
  >
    Cargando... 
  </div>
  <div
    className="card-grid"
  />
</Fragment>
`;
